storm-core/src/jvm/org/apache/storm/daemon/StormCommon.java

```diff
@@ -0,0 +1,604 @@
+
+        List<StormTopology._Fields> spoutFields = Arrays.asList(Thrift.getSpoutFields());
+        for (StormTopology._Fields field : spoutFields) {
+            Map<String, Object> spoutComponents = (Map<String, Object>) topology.getFieldValue(field);
+            if (spoutComponents != null) {
+                for (Object obj : spoutComponents.values()) {
+                    ComponentCommon common = getComponentCommon(obj);
+                    if (isEmptyInputs(common) == false) {
+                        throw new InvalidTopologyException("May not declare inputs for a spout");
+                    }
+                }
+            }
+        }
+
+        Map<String, Object> componentMap = allComponents(topology);
+        for (Object componentObj : componentMap.values()) {
+            Map conf = componentConf(componentObj);
+            ComponentCommon common = getComponentCommon(componentObj);
+            if (common != null) {
+                int parallelismHintNum = Thrift.getParallelismHint(common);
+                Integer taskNum = Utils.getInt(conf.get(Config.TOPOLOGY_TASKS), 0);
+                if (taskNum > 0 && parallelismHintNum <= 0) {
+                    throw new InvalidTopologyException("Number of executors must be greater than 0 when number of tasks is greater than 0");
+                }
+            }
+        }
+    }
+
+    private static Set<String> getStreamOutputFields(Map<String, StreamInfo> streams) {
+        Set<String> outputFields = new HashSet<String>();
+        if (streams != null) {
+            for (StreamInfo streamInfo : streams.values()) {
+                outputFields.addAll(streamInfo.get_output_fields());
+            }
+        }
+        return outputFields;
+    }
+
+    public static void validateStructure(StormTopology topology) throws InvalidTopologyException {
+        Map<String, Object> componentMap = allComponents(topology);
+        for (Map.Entry<String, Object> entry : componentMap.entrySet()) {
+            String componentId = entry.getKey();
+            ComponentCommon common = getComponentCommon(entry.getValue());
+            if (common != null) {
+                Map<GlobalStreamId, Grouping> inputs = common.get_inputs();
+                for (Map.Entry<GlobalStreamId, Grouping> input : inputs.entrySet()) {
+                    String sourceStreamId = input.getKey().get_streamId();
+                    String sourceComponentId = input.getKey().get_componentId();
+                    if(componentMap.keySet().contains(sourceComponentId) == false) {
+                        throw new InvalidTopologyException("Component: [" + componentId + "] subscribes from non-existent component [" + sourceComponentId + "]");
+                    }
+
+                    ComponentCommon sourceComponent = getComponentCommon(componentMap.get(sourceComponentId));
+                    if (sourceComponent == null || sourceComponent.get_streams().containsKey(sourceStreamId) == false) {
```

Comment:
The null check for sourceComponent was not in the original code, and if it were null it would not mean that we subscribed to a non-existent stream.  It would mean that our topology builder didn't do its job correctly.